src/id-data-controls.tsx

```diff
@@ -0,0 +1,293 @@
+        });
+        savedIdData.getSignature().then((signature) => signature && setSignature(signature));
+    }, [savedIdData]);
+
+    useEffect(() => {
+        resetSavedIdData();
+    }, [resetSavedIdData]);
+
+    useEffect(() => {
+        savedIdData.storeArray(fieldsArrayFromFixedData(fixedIdData));
+    }, [savedIdData, fixedIdData]);
+
+    useEffect(() => {
+        savedIdData.storeArray(customIdData, false);
+    }, [savedIdData, customIdData]);
+
+    const handleFixedChange = (type: 'name' | 'birthdate' | 'email' | 'address', value: string | Address) => {
+        setFixedIdData(
+            produce((data) => {
+                if (!isAddress(value)) {
+                    if (type !== 'address') data[type] = value;
+                } else {
+                    if (type === 'address') data[type] = { ...value, primary: true };
+                }
+            })
+        );
+    };
+
+    if (Privacy.isAllowed(PRIVACY_ACTIONS.SAVE_ID_DATA)) {
+        return (
+            <div id="id-data-controls-container" className="narrow-page">
+                <DynamicInputContainer
+                    key="id-data-controls"
+                    id="id-data-controls"
+                    onAddField={(new_field) =>
+                        setCustomIdData(
+                            produce((fields) => {
+                                const last_index =
+                                    customIdData
+                                        .map((f) =>
+                                            parseInt(
+                                                f.desc.match(
+                                                    new RegExp(`^${t('custom-input-desc', 'id-data-controls')} (\\d+)$`)
+                                                )?.[1] ?? '',
+                                                10
+                                            )
+                                        )
+                                        // remove the undefined elements
+                                        .filter((d) => d)
+                                        .sort()
+                                        .pop() || 0;
+                                fields.push({
+                                    ...new_field,
+                                    desc:
+                                        new_field.desc ||
+                                        `${t('custom-input-desc', 'id-data-controls')} ${last_index + 1}`,
+                                });
+                            })
+                        )
+                    }
+                    onRemoveField={(idx) => {
+                        if (!customIdData[idx])
+                            throw new ErrorException(
+                                'Index out of bounds for customIdData (in removeField).',
+                                {
+                                    index: idx,
+                                    customIdData: { ...customIdData },
```

Comment:
I don't understand why you'd need to copy these but I'm sure you have your reasons.